### PR TITLE
Retry importing server certificate if private key is not accessible

### PR DESCRIPTION
**Pull request by vipeller**
This is a workaround for https://github.com/Azure/iotedge/issues/5087

When EdgeHub starts up and imports the certificates with their private keys, sometimes the private key does not get associated with the certificate. As a result, using the certificate object and accessing the private key throws an exception.
Because this certificate is handed over to Kestrel to provide a TLS connection, Kestrel tries to access the private key and throws an exception every time a client tries to connect. As a result, no client can connect to EdgeHub and it never recovers from this error.

We could not find the root cause that why the import fails sometimes and other times not, also it is not clear why only on windows. However, what we found is that retrying the import (with the same certificate and key files) usually works (always worked in the tests, but we cannot claim that it always will).

As a workaround, during startup, edge hub tries to access the private key to see if the association between the certificate and the key is working. If an error occurs, it re-imports the certificate, and it tries at max 10 times before stopping. Stop is needed because it is possible that the certificate files themself not correct and we want to avoid endless tries.

The diff tool shows excessive change, however only a few lines was added at the beginning of the file and around line 100.

### CODE REVIEW

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs [hunk 1]
```diff
@@ -33,64 +36,92 @@ public class EdgeHubCertificates
...
+                {
+                    throw new InvalidOperationException("Edge Hub certificate files incorrectly configured");
+                }
+
+                // On Windows, from time to time the private key cannot be accessed and kestrel fails accepting connections
+                // without the private key. Testing the private key here and if it fails, retry the entire cert import.
+                // The second try usually works.
+                try
+                {
+                    _ = result.ServerCertificate.PrivateKey;
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs:109]: So the only change is to check if the ServerCertificate has a private key?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs [hunk 2]
```diff
@@ -33,64 +36,92 @@ public class EdgeHubCertificates
...
-            {
-                throw new InvalidOperationException("Edge Hub certificate files incorrectly configured");
+                    var workloadUri = new Uri(configuration.GetValue<string>(Constants.ConfigKey.WorkloadUri));
+                    string edgeHubHostname = configuration.GetValue<string>(Constants.ConfigKey.EdgeDeviceHostName);
+                    string moduleId = configuration.GetValue<string>(Constants.ConfigKey.ModuleId);
+                    string generationId = configuration.GetValue<string>(Constants.ConfigKey.ModuleGenerationId);
+                    string edgeletApiVersion = configuration.GetValue<string>(Constants.ConfigKey.WorkloadAPiVersion);
+                    DateTime expiration = DateTime.UtcNow.AddDays(Constants.CertificateValidityDays);
+
+                    certificates = await CertificateHelper.GetServerCertificatesFromEdgelet(workloadUri, edgeletApiVersion, Constants.WorkloadApiVersion, moduleId, generationId, edgeHubHostname, expiration);
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs:66]: Can we instead move this retry logic to CertificateHelper.GetServerCertificatesFromEdgelet? If the server cert it gets has no private key, then we try to get the cert again.

**Review by varunpuranik — COMMENTED** [review]
Have we been able to repro this at all? If so, what do edged logs look like? Anything interesting there? 
I am really curious to know why this is happening.. and seems to be in 1.1

**Comment by vipeller** [comment]
yes, I could repro it in a VM and also saw the message in the logs about retrying (then edgehub worked). I have not checked edged, however the way anca tested it, she moved the import-code into the code of the temperature sensor, using bouncy castle to create the pkcs12 content and import it with the X509Certificate class (just like edgehub does), she also saw this error (without edgehub or edge involved). It is eighter a bug in bouncy castle creating a bad pkcs12 file (however I don't see how X509Certificate2 does not complain about the bad file), or something is wrong with the X509Certificate2 importing in .net/windows.
According to some descriptions, in order to use the private key with the certificate, it must be moved into some certificate store (that the X509Certificate.Import() does), otherwise the certificate will not be able to use the private key and it throws the exact same error message. I figured (although not too sure) that this step fails somehow sometimes on windows.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs (hunk 2 above)
Comment (vipeller) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs:66]: the 'GetServerCertificateAndChainFromFile' call is also affected. I can move the logic into 'ParseCertificateAndKey()' I think

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs (hunk 1 above)
Comment (vipeller) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs:109]: yes, this is what makes it break the way kestrel would do. As I picture, during connection kestrel does the same to prove the client that is has the private key for its certificate it sent to the client, but fails to obtain it and throws the message we see in the logs.
This call just does it much earlier, so if it would break during connection, we see it now and try to reimport (which magically works somehow)

**Comment by varunpuranik** [comment]
> yes, I could repro it in a VM and also saw the message in the logs about retrying (then edgehub worked). I have not checked edged, however the way anca tested it, she moved the import-code into the code of the temperature sensor, using bouncy castle to create the pkcs12 content and import it with the X509Certificate class (just like edgehub does), she also saw this error (without edgehub or edge involved). It is eighter a bug in bouncy castle creating a bad pkcs12 file (however I don't see how X509Certificate2 does not complain about the bad file), or something is wrong with the X509Certificate2 importing in .net/windows. According to some descriptions, in order to use the private key with the certificate, it must be moved into some certificate store (that the X509Certificate.Import() does), otherwise the certificate will not be able to use the private key and it throws the exact same error message. I figured (although not too sure) that this step fails somehow sometimes on windows.

I see, so you are saying this is not an issue with edged but just with bouncy-castle? Interesting.. can we confirm this? And if so, then maybe we can just try to load the same bytes returned by edged into the object multiple times?

**Comment by vipeller** [comment]
either bouncy castle when it creates the pcks12 content or X509Certificate.Import() (when it imports the pkcs12 content). I would bet the second, because that case seems odd that bouncy castle has a bug, it generates a broken pkcs12 sometimes, but even that it is broken, good enough for X509Certificate to import it without any exception.  
I think the byte content from edged could be reused (anca tried this with hardcoded certificate/private key, which simulated the case when we already received the cert/key from edged)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs [hunk 3]
```diff
@@ -33,7 +31,9 @@ public class EdgeHubCertificates
         public static async Task<EdgeHubCertificates> LoadAsync(IConfigurationRoot configuration, ILogger logger)
         {
             Preconditions.CheckNotNull(configuration, nameof(configuration));
+
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs:34]: Nit: I see some odd formatting changes in this file.. can you pls revert them?

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs [hunk 4]
```diff
@@ -373,67 +379,103 @@ internal static X509Certificate2 ParseManifestTrustedBundleCerts(string manifest
...
+                    }
+                    catch
+                    {
+                        // swallow
+                    }
 
-                var cert = new X509Certificate2(p12File.ToArray());
-                return (cert, certsChain);
+                    logger?.LogWarning("Error importing certificate, retrying");
+                    Thread.Sleep(TimeSpan.FromSeconds(1)); // Do not spam the log
```
Comment (varunpuranik) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs:474]: I am fine with it... Why do we need to sleep here? Is this related to some timing issue?

**Review by varunpuranik — APPROVED** [review]

**Comment by vipeller** [comment]
recreated for 1.1: https://github.com/Azure/iotedge/pull/5930

**Comment by vipeller** [comment]
recreated for 1.1 https://github.com/Azure/iotedge/pull/5930
